### PR TITLE
Switch `Stream` back to having an `AttributeFilter` field and add `New*Filter` functions

### LINKED ISSUE #4156: View attribute filtering is not compliant with the specification
The specification states the stream configuration of a view must have an allow-list of attribute keys that it will use to filter out attributes:

>   * A list of `attribute keys` (optional). If provided, the attributes that are not in the list will be ignored. If not provided, all the attribute keys will be used by default (TODO: once the Hint API is available, the default behavior should respect the Hint if it is available).

However, the opentelemetry-go implementation uses an [`attribute.Filter`](https://github.com/open-telemetry/opentelemetry-go/blob/e2f8ecae74aba268326d2fbf52c6ccce21e56698/attribute/set.go#L47).

The non-compliance of the implementation needs to be addressed

### Proposal

Update the specification to allow something like the `attribute.Filter` to be used for this functionality.

Having a list of attribute keys that are allowed is in the stream configuration limits the usefulness of attribute filtering. It only allows an allow-list type filtering for static pre-determined attribute keys. Having something allowed like the `attribute.Filter` allows for deny-lists, filtering on values, fuzzy filtering algorithms, probabilistic filtering ... It seems like users would benefit by keeping the `attribute.Filter` here instead of replacing it with a `[]attribute.Key` field.

### Alternatives

Replace the `attribute.Filter` type with `[]attribute.Key` in the `Stream.AttributeFilter` field. Use these keys to filter attributes.

### PR DESCRIPTION
**Pull request by MrAlias**
The existing fix to #4156 (#4288) introduced a change that limited the functionality of the SDK. Specifically it prevents the solution of the type of problems found in #3765. This does the following to address #4156:

- Revert "Replace `Stream.AttributeFilter` with `AllowAttributeKeys` (#4288)"
- Add `NewAllowKeysFilter` and `NewDenyKeysFilter` to `otel/attribute` to provide easy allow-keys and deny-keys filter creation.

### CODE REVIEW

**Comment by dashpole** [comment]
Have we gotten this OK'd with a TC reviewer as being spec compliant?

**Comment by MrAlias** [comment]
> Have we gotten this OK'd with a TC reviewer as being spec compliant?

Not yet. Still working on it.

**Review by dashpole — APPROVED** [review]
LGTM, as long as it complies with the spec.

File: sdk/metric/benchmark_test.go [hunk 1]
```diff
@@ -42,7 +42,9 @@ var viewBenchmarks = []struct {
 		"AttrFilterView",
 		[]View{NewView(
 			Instrument{Name: "*"},
-			Stream{AllowAttributeKeys: []attribute.Key{"K"}},
+			Stream{AttributeFilter: func(kv attribute.KeyValue) bool {
+				return kv.Key == attribute.Key("K")
+			}},
```
Comment (MadVikingGod) [inline sdk/metric/benchmark_test.go:47]: While this is probably more optimized, how does it compare with using a `NewAllowKeysFilter("K")`?

If they are in the same ballpark, I recommend using the functions we created for this.

File: sdk/metric/meter_test.go [hunk 2]
```diff
@@ -1518,7 +1518,9 @@ func testAttributeFilter(temporality metricdata.Temporality) func(*testing.T) {
 					WithReader(rdr),
 					WithView(NewView(
 						Instrument{Name: "*"},
-						Stream{AllowAttributeKeys: []attribute.Key{"foo"}},
+						Stream{AttributeFilter: func(kv attribute.KeyValue) bool {
+							return kv.Key == attribute.Key("foo")
+						}},
```
Comment (MadVikingGod) [inline sdk/metric/meter_test.go:1523]: This should use the `NewAllowKeysFilter()` to show how these components fit together.

File: sdk/metric/pipeline.go [hunk 3]
```diff
@@ -351,8 +351,8 @@ func (i *inserter[N]) cachedAggregator(scope instrumentation.Scope, kind Instrum
 		b := aggregate.Builder[N]{
 			Temporality: i.pipeline.reader.temporality(kind),
 		}
-		if len(stream.AllowAttributeKeys) > 0 {
-			b.Filter = stream.attributeFilter()
+		if stream.AttributeFilter != nil {
+			b.Filter = stream.AttributeFilter
 		}
```
Comment (MadVikingGod) [inline sdk/metric/pipeline.go:356]: Can this be just `b.Filter = stream.AttributeFilter`?  If it's nil it will stay nil, if not, we get a filter.

**Review by MadVikingGod — COMMENTED** [review]
If there is TC sign-off, then I Approve this.

File: sdk/metric/benchmark_test.go (hunk 1 above)
Comment (MrAlias) [reply sdk/metric/benchmark_test.go:47]: Sure, sounds good. I had just reverted, but using the new func sounds good to me.

**Comment by MrAlias** [comment]
After a slack conversation with @jack-berg he pointed out that @reyang, @jsuereth, and @jmacd will have more context here. I have created https://github.com/open-telemetry/opentelemetry-specification/issues/3664 and pinged them in the issue.

**Review by hanyuancheung — APPROVED** [review]

**Review by MadVikingGod — APPROVED** [review]
I approve this with the assumption that this is for beta release.  The Spec compliance is still in process.

**Review by dashpole — APPROVED** [review]

**Pull request merged**